driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,174 @@
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
+                Ok(()) | Err(RecvTimeoutError::Timeout) => {
+                    // Make sure we don't hold the mutex while this blocking call happens.
+                    match price_source.get_prices(&tokens) {
+                        Ok(prices) => price_map_.lock().unwrap().extend(prices),
+                        Err(err) => log::warn!("price_source::get_prices failed: {}", err),
+                    }
+                }
+                // The owning struct has been dropped.
+                Err(RecvTimeoutError::Disconnected) => break,
+            }
+
+            #[cfg(test)]
+            let _ = test_sender.send(());
+        });
+
+        // Update the prices immediately.
+        sender.send(()).unwrap();
+        (
+            Self {
+                price_map,
+                _channel_to_thread: sender,
+                #[cfg(test)]
+                test_receiver,
+            },
+            join_handle,
+        )
+    }
+}
+
+impl PriceSource for ThreadedPriceSource {
+    /// Non blocking.
+    /// Infallible.
+    fn get_prices(&self, _tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(self.price_map.lock().unwrap().clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::price_source::MockPriceSource;
+    use super::*;
+    use std::sync::atomic;
+    use std::time::Instant;
+
+    const ORDERING: atomic::Ordering = atomic::Ordering::SeqCst;
+    const THREAD_TIMEOUT: Duration = Duration::from_secs(1);
+
+    lazy_static::lazy_static! {
+        static ref TOKENS: [Token; 3] = [
+            Token::new(0, "0", 0),
+            Token::new(1, "1", 1),
+            Token::new(2, "2", 2),
+        ];
+    }
+
+    /// Drops tps and joins the handle which ensures that the thread exits and
+    /// hasn't panicked. This is important because the mockall expectation
+    /// panics happen in the thread which would otherwise let the tests pass
+    /// when an expectation is violated.
+    fn join(tps: ThreadedPriceSource, handle: JoinHandle<()>) {
+        fn extract_test_receiver(tps: ThreadedPriceSource) -> Receiver<()> {
+            tps.test_receiver
+        }
+        let test_receiver = extract_test_receiver(tps);
+        let deadline = Instant::now() + THREAD_TIMEOUT;
+        loop {
+            match test_receiver.recv_timeout(THREAD_TIMEOUT) {
+                // The sender side of `test_receiver` has been dropped which
+                // only happens when the thread exits.
+                Err(RecvTimeoutError::Disconnected) => break,
+                // There is still a queued messages on the channel.
+                Ok(_) if Instant::now() <= deadline => continue,
```

Comment:
As you said we can still run into deadlock here if we happened to produce more messages than we can consume (e.g. if we produced fetched prices every 1ns or so).

Since multithreading is so hard, it might be nicer to inject a "periodic executor" into the threaded price source which can be instantiated with a closure and implements `execute_now()`. In the tests you could implement a synchronous version of it and for prod you use the thread + channel version.

In order to test the right drop behavior (which is a nice test) you could then keep the back-channel without the need of sending a message after each execution which should make it not stallable.

Reply:
There is no deadlock here because we are testing against deadline. If there are still messages after the deadline we panic.

Technically there is such a deadlock in the function below at "Clear previous messages". I changed this to also have a timeout.

With regards to the "executer": I specifically want to test the real threading behavior, not a synchronous test executor. Additionally consider that we can also have a deadlock in single threaded with an accidental infinite loop and we don't wrap all our other tests in timeouts either.

Reply:
I see your point about deadlocking on infinity loops. I don't like to use infinity loops (I remember a comment on a previous PR), unless e.g. in a top level thread that is not supposed to terminate or when it's absolutely obvious the logic will exit.
The advantage about using your own executor is that you can completely control the threading behaviour (things like when it gets scheduled compared to logic happening on another thread). In my opinion the actual thread::spawn is likely not test worthy. 
With the current approach we are having a potentially non-deterministic test, and understanding the test setup for a future maintainer is quite hard.